library/ssl_tls.c

```diff
@@ -1921,6 +1921,54 @@ int mbedtls_ssl_set_hs_ecjpake_password( mbedtls_ssl_context *ssl,
 
     return( 0 );
 }
+
+int mbedtls_ssl_set_hs_ecjpake_password_opaque( mbedtls_ssl_context *ssl,
+                                                mbedtls_svc_key_id_t pwd )
+{
+    psa_pake_cipher_suite_t cipher_suite = psa_pake_cipher_suite_init();
+    psa_pake_role_t psa_role;
+    psa_status_t status;
+
+    if( ssl->handshake == NULL || ssl->conf == NULL )
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+    if( ssl->conf->endpoint == MBEDTLS_SSL_IS_SERVER )
+        psa_role = PSA_PAKE_ROLE_SERVER;
+    else
+        psa_role = PSA_PAKE_ROLE_CLIENT;
+
+    if( mbedtls_svc_key_id_is_null( pwd ) )
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    ssl->handshake->psa_pake_password = pwd;
+
+    psa_pake_cs_set_algorithm( &cipher_suite, PSA_ALG_JPAKE );
+    psa_pake_cs_set_primitive( &cipher_suite,
+                               PSA_PAKE_PRIMITIVE( PSA_PAKE_PRIMITIVE_TYPE_ECC,
+                                                   PSA_ECC_FAMILY_SECP_R1,
+                                                   256) );
+    psa_pake_cs_set_hash( &cipher_suite, PSA_ALG_SHA_256 );
+
+    status = psa_pake_setup( &ssl->handshake->psa_pake_ctx, &cipher_suite );
+    if( status != PSA_SUCCESS )
+        goto error;
```

Comment:
Sorry but I'm afraid I didn't get the point here. The documentation of `psa_pake_setup` says that:

> If an error occurs at any step after a call to psa_pake_setup(), the operation will need to be reset by a call to psa_pake_abort()

This seems to be the case in `mbedtls_ssl_set_hs_ecjpake_password_opaque()`. What am I missing?

Reply:
> If an error occurs at any step **after** a call to psa_pake_setup() [...]

So an error in `psa_pake_setup` itself does not require calling `psa_pake_abort` afterwards :) 
You can see that it's already called here: https://github.com/Mbed-TLS/mbedtls/blob/90af1a10abc187186ca3bb48200ede1a1f429042/library/psa_crypto_pake.c#L251

Reply:
Oh gosh, albeit I was the one that did this change, I didn't remember it and I don't have the current branch, so I got confused. Sorry! I'm going to fix that

Reply:
Well, again we shouldn't be looking at any `psa_xxx.c` file at this point, only the spec. I'm not 100% convinced by your interpretation of the spec here, @AndrzejKurek, so I've asked on slack what others think. IMO if `psa_pake_setup()` itself returns an error, that's an error that happens after the call to `psa_pake_setup()` since the order is (1) we call it, then (2) it returns an error. But I see the logic in you interpretation and I think the wording in the spec could be clearer.

Reply:
Ok, Gilles's interpretation is that we should call `abort()` when `psa_pake_setup()` itself fails if we want to be portable (and we do). Also, it's always safe to call `abort()` even when it's not needed, while the opposite isn't, so let's call it.

Reply:
It would be great to fix the spec to follow the same convention for `xxx_setup` operations, as I don't see a reason to not do so. But sure, I can agree that it can be interpreted the other way around too.